File: client/src/lib/corpusCache.ts
import { invoke } from '@tauri-apps/api/core';

// Corpus SWR cache. Backed by a plain JSON file at %APPDATA%/whatsub/
// corpus_cache.json (via the corpus_cache_load/save Rust commands) so it lives
// in the same unified dir as settings/license/auth — NOT the tauri-plugin-store
// bundle-identifier dir (com.whatsub.app/) it used to use. The Rust side
// migrates the legacy plugin-store file on first load.
//
// The whole cache is a small {key: value} map; we hold it in memory after a
// one-time load and persist the full map on every mutation (file is a few KB).

let _cache: Record<string, unknown> | null = null;
let _loading: Promise<Record<string, unknown>> | null = null;

async function load(): Promise<Record<string, unknown>> {
  if (_cache) return _cache;
  if (!_loading) {
    _loading = (async () => {
      try {
        const raw = await invoke<string>('corpus_cache_load');
        _cache = raw ? (JSON.parse(raw) as Record<string, unknown>) : {};
      } catch {
        _cache = {};
      }
      return _cache;
    })();
  }
  return _loading;
}

async function persist(): Promise<void> {
  if (!_cache) return;
  try {
    await invoke('corpus_cache_save', { contents: JSON.stringify(_cache) });
  } catch {
    /* best-effort cache write — a failed persist just means a stale-cache
       refetch next launch, never a user-visible error */
  }
}

export type VersionKey = 'mineVersion' | 'publicVersion';

export async function getCachedVersion(key: VersionKey): Promise<number> {
  const c = await load();
  const v = c[key];
  return typeof v === 'number' ? v : 0;
}

export async function setCachedVersion(key: VersionKey, v: number): Promise<void> {
  const c = await load();
  c[key] = v;
  await persist();
}

export async function getCachedData<T>(key: string): Promise<T | null> {
  const c = await load();
  return (c[key] as T | undefined) ?? null;
}

export async function setCachedData<T>(key: string, data: T): Promise<void> {
  const c = await load();
  c[key] = data;
  await persist();
}

export async function invalidate(...keys: string[]): Promise<void> {
  const c = await load();
  for (const k of keys) delete c[k];
  await persist();
}

export async function invalidateAll(): Promise<void> {
  _cache = {};
  await persist();
}

File: client/src-tauri/src/commands/corpus_cache.rs
//! Corpus SWR cache persistence.
//!
//! Stores the corpus version/list cache as a plain JSON `{key: value}` map at
//! `%APPDATA%/whatsub/corpus_cache.json` — the unified hand-rolled app-data dir
//! (see core::paths), NOT the tauri-plugin-store bundle-identifier dir
//! (`com.whatsub.app/`) it used to live in. `corpus_cache_load` migrates the
//! legacy plugin-store file on first read so the cache survives the move.

use crate::core::paths;
use crate::error::{AppError, AppResult};
use tauri::{AppHandle, Manager};

#[tauri::command]
pub fn corpus_cache_load(app: AppHandle) -> AppResult<String> {
    let path = paths::corpus_cache_path().map_err(AppError::Other)?;
    if let Ok(s) = std::fs::read_to_string(&path) {
        return Ok(s);
    }
    // One-time migration from the old tauri-plugin-store location
    // (<app_data_dir-by-identifier>/corpus_cache.json). Same plain JSON shape.
    if let Ok(dir) = app.path().app_data_dir() {
        let legacy = dir.join("corpus_cache.json");
        if let Ok(s) = std::fs::read_to_string(&legacy) {
            if let Some(parent) = path.parent() {
                let _ = std::fs::create_dir_all(parent);
            }
            let _ = std::fs::write(&path, &s);
            let _ = std::fs::remove_file(&legacy);
            return Ok(s);
        }
    }
    Ok("{}".to_string())
}

#[tauri::command]
pub fn corpus_cache_save(contents: String) -> AppResult<()> {
    let path = paths::corpus_cache_path().map_err(AppError::Other)?;
    if let Some(parent) = path.parent() {
        std::fs::create_dir_all(parent)?;
    }
    std::fs::write(&path, contents)?;
    Ok(())
}
